### PR TITLE
Confusion self-hit doesn't trigger Sitrus + super sitrus Berries

### PR DESCRIPTION
**Pull request by QuiteQuiet**
Mentioned in https://github.com/Zarel/Pokemon-Showdown/issues/2367#issuecomment-327332860

Not sure if this is the best way to implement this, but I can't think of anything else.

@Marty or @DaWoblefet can either of you confirm if Oran Berry/Berry Juice/Stat-boosting berries also does this or if it's only the healing berries. Also is this new in Gen 7, or was this also the case before?

### CODE REVIEW

**Comment by DaWoblefet** [comment]
I can confirm this has been the case since Gen 5 (not sure about prior). I'll check the others today.

**Comment by Marty-D** [comment]
The thing about `Update` is it triggers almost all the time. Even if you prevent these Berries from activating on the self-hit from confusion, they'll just activate immediately after you next use a move or confusion ends. The Berries shouldn't activate until your HP is next updated by an attack or something.

Also re: the test, the only reason it passes is because Swagger can miss, Deoxys might not hit itself in confusion, and even if it does, it knocks itself out in one hit at +2 so the Sitrus Berry is never used.

**Comment by QuiteQuiet** [comment]
Does this work better then? I can't think of anything else beside doing something with damage events.

>Also re: the test, the only reason it passes is because Swagger can miss, Deoxys might not hit itself in confusion, and even if it does, it knocks itself out in one hit at +2 so the Sitrus Berry is never used.

Will this be enough? Test PRNG will hit + self-hit in confusion on the first turn, so it shouldn't be able to pass due to luck.
0 Atk Deoxys-Attack Struggle vs. 0 HP / 0 Def Deoxys-Attack: 150-177 (62.2 - 73.4%) -- guaranteed 2HKO

**Comment by Marty-D** [comment]
The `confused` check for Figy and friends seems to be the opposite of what you want but otherwise this looks like a great solution! 👍

**Comment by QuiteQuiet** [comment]
rip did

Still wondering about Oran Berry/Berry Juice/etc. though.

**Comment by Zarel** [comment]
So what's wrong with listening to damage events?

**Comment by QuiteQuiet** [comment]
If you mean `onDamage`:

     1) Emergency Exit should not request switch-out if first healed by berry:
    
         AssertionError [ERR_ASSERTION]: 'switch' === 'move'
         + expected - actual
    
         -switch
         +move
    
     2) Pickup should pick up a consumed item:
        AssertionError [ERR_ASSERTION]: Pick Up should retrieve consumed Sitrus Berry
    
     3) Sitrus Berry should heal 25% hp when consumed:
        AssertionError [ERR_ASSERTION]: Expected 'holdsItem' assertion to fail.
    
     4) Heal Block [Gen 4] should allow HP recovery items to activate:
    
         AssertionError [ERR_ASSERTION]: 'sitrusberry' === ''
         + expected - actual
    
         -sitrusberry

And if you meant `onAfterDamage`, this makes direct damage like Substitute (and various other things that uses `directDamage`) not trigger the berries at all. I don't know of any other damage event that could be reasonable, but neither of those work without breaking something different.

**Comment by Zarel** [comment]
We could fix `onAfterDamage`...

**Comment by QuiteQuiet** [comment]
Would it not be better to make `directDamage` trigger `onAfterDamage` (unless this was what you meant)? Or is there a reason why it doesn't at the moment.

**Comment by Zarel** [comment]
Yes, that's what I was proposing.

**Comment by DaWoblefet** [comment]
> @Marty or @DaWoblefet can either of you confirm if Oran Berry/Berry Juice/Stat-boosting berries also does this or if it's only the healing berries

Here's my testing (if you need the video for confirmation / how the interaction works, let me know).

- Oran Berry - does not activate after confusion
- Berry Juice - does not activate after confusion
- Stat pinch Berries - does not activate after confusion
- Berserk - does not activate after confusion
- Emergency Exit / Wimp Out - does not activate after confusion. Subsequent attacks that would damage the Emergency Exit / Wimp Out Pokemon below 50% do not activate the ability.

- Zen Mode - activates normally
- Schooling - activates normally
- Shields Down - activates normally
- Soul Heart - activates normally
- Defeatist - halves the user's stats normally

**Comment by QuiteQuiet** [comment]
Doing that, the remaining problem from what I can see is that non-damage events (like Harvest harvesting a berry) won't then continue to trigger the Berry, which I think is the intended mechanic. There may be a way to work around that with some other event maybe?

**Comment by Zarel** [comment]
Attach it to `onStart` too?

**Comment by QuiteQuiet** [comment]
Uh this doesn't work as it did before. There's some problems with the order that item effects happen (Recycle eat the berry before notifying that that it got one). I'll try to see if I can fix this somehow.

**Comment by QuiteQuiet** [comment]
Okay, this might work then. I'm not entirely sure on the activation order after Recycle Air Balloon, but otherwise this should be good.

**Comment by Marty-D** [comment]
There's at least one regression with this approach. If you're facing something with Unnerve and it knocks you into range for your Berry to activate, then it switches out, your Berry wouldn't activate immediately anymore. (In fact, it should activate even more immediately than it does currently since switching to another Unnerve Pokemon stops the Berry from activating, which is itself a bug.)

**Comment by QuiteQuiet** [comment]
That's getting to a quite a lot of different events to consider... How many other likely regressions will come from this (Does Magic Room running out make a Pokemon eat a Berry straight away if they are in range for example?).

I guess an option would be to activate the `onStart` from the end events of the respective source that blocks items like this, but from what I see these events run before the switch actually happen, which would lead to berries being eaten too early instead...

    onEnd: function (pokemon) {
    	for (let i = 0; i < pokemon.side.foe.active.length; i++) {
    		let active = pokemon.side.foe.active[i];
    		let item = active.getItem();
    		this.singleEvent('Start', item, active.itemData, active, pokemon, item);
    	}
    },

**Comment by MacChaeger** [comment]
I wonder what happens if a Pokemon knocks itself below half through confusion, keeping its berry, and then something with Unnerve switches in and out, or if the confusion hit happens while the berry is being negated through Unnerve, and _then_ the Unnerve mon switches.

**Comment by Zarel** [comment]
At this point, I think the approach of "Update event, don't activate if previous event was confusion" might be the best approach.

But I'm starting to wonder how exactly _Game Freak_ managed this bug. What happens if you hurt yourself in confusion, and then an opposing Unnerve user switches out?

**Comment by QuiteQuiet** [comment]
I'm honestly not sure what the best way around this is, and there's some interactions that I can't find information on (Magic Room expire after you hit yourself in confusion, do you eat the berry?, or Zarel's Unnerve situation).

**Comment by Marty-D** [comment]
>(Magic Room expire after you hit yourself in confusion, do you eat the berry?)

>I wonder what happens if a Pokemon knocks itself below half through confusion, keeping its berry, and then something with Unnerve switches in and out before the Pokemon with the uneaten berry takes damage, or if the confusion hit happens while the berry is being negated through Unnerve, and then the Unnerve mon switches.

All of these scenarios trigger the Berry.

**Comment by QuiteQuiet** [comment]
This should not cause any regressions (finally), but it doesn't fix the Unnerve into Unnerve problem Marty brought up earlier.

File: data/items.js [hunk 1]
```diff
@@ -396,7 +398,7 @@ exports.BattleItems = {
 			basePower: 30,
 		},
 		onUpdate: function (pokemon) {
-			if (pokemon.hp <= pokemon.maxhp / 2) {
+			if (pokemon.hp <= pokemon.maxhp / 2 && pokemon.lastEffect && pokemon.lastEffect.id !== 'confused') {
```
Comment (Marty-D) [inline data/items.js]: Could `lastEffect` ever be `null` here? Like say you just switched in and your HP is already in range. Maybe return early like you've done for the rest of the Berry checks just to be extra safe.

Comment (QuiteQuiet) [reply data/items.js]: It could, probably, so yeah probably better to move it outside.

**Comment by QuiteQuiet** [comment]
@Marty-D After some experimentation regarding
>In fact, it should activate even more immediately than it does currently since switching to another Unnerve Pokemon stops the Berry from activating, which is itself a bug.

it seems as the below would "fix" it, though I don't know how cart activates the berry (in-between switches when nothing is on the field/before the pokeball or message for the new pokemon is shown/before the unnerve activation message) so I cannot claim this is an actual fix to the issue.
```diff
		onPreStart: function (pokemon) {
			this.add('-ability', pokemon, 'Unnerve', pokemon.side.foe);
		},
-		onFoeTryEatItem: false,
+		onFoeTryEatItem: function () {
+			return !this.effectData.target.isStarted;
+		},
		id: "unnerve",
		name: "Unnerve",
		rating: 1.5,
```

**Comment by Zarel** [comment]
It sounds like the best solution might be a `StartUpdate` event that triggers on switch-in and also when Magic Room or some other effect suppressing items ends.

**Comment by QuiteQuiet** [comment]
I don't think I know how the event system PS uses well enough to even know where to start in implementing something like that.

**Comment by Marty-D** [comment]
>in-between switches when nothing is on the field/before the pokeball or message for the new pokemon is shown

It is this, yeah. No need to have it fixed right now since I think it's beyond the scope of this PR, I just mentioned it to clarify what I meant by "immediately" in that case.

**Comment by QuiteQuiet** [comment]
So, what happens now?

**Comment by Zarel** [comment]
I'm not sure I like the idea of adding a new member `.lastEffect` to the Pokémon object, purely to work around a lack of familiarity with the event system.

**Comment by MacChaeger** [comment]
What if berries were refactored to only activate onAfterDamage and onStart, and effects that negate berries and other items trigger the onStart effect of all affected items when they end, like QuiteQuiet suggested earlier? That would probably end up causing a regression with Air Balloon (do something with onRestart?) but it sounds to me like that would be closer to how Game Freak implements berries in-game.

**Comment by QuiteQuiet** [comment]
Realistically, this is the best I can manage. There's nothing that can keep the `onUpdate` events intact without adding a new tracker since Update events have no idea what caused the last damage.

It's basically what was before, plus the Unnerve/Magic Room stuff working properly...

**Comment by Marty-D** [comment]
Here are some tests I did in Gen 6 along with UPC's "after an attack" ordering, which I've amended using my results. Make of this what you will.

```
Flare Blitz -> burn -> recoil -> target's Sitrus
Knock Off -> Rocky Helmet damage -> target faints -> Knock Off effect
Knock Off -> Rocky Helmet damage -> user's Sitrus -> Knock Off effect
Knock Off damage brings target in range for Sitrus activation -> Knock Off effect
Knock Off -> target faints -> Destiny Bond knocks out user
Poison Touch Thunder Punch -> paralysis -> target's Lum Berry -> poison
Flare Blitz -> Rough Skin -> recoil -> target thaws -> target's Kee
Dragon Tail -> target is forced out before Kee Berry activates
Dragon Tail -> target is forced out before Sitrus activates
Rapid Spin -> user is freed from stuff -> target's Sitrus
Flame Burst -> bursting flame -> target thaws -> target's Sitrus
Poison Touch contact recoil move on a Pokemon using Rage -> rage is building -> poison -> target's Lum -> recoil
Flare Blitz -> recoil -> user's Sitrus -> target's Kee
Flare Blitz -> Rocky Helmet -> user's Sitrus -> recoil
Flare Blitz -> recoil -> user's Sitrus -> target thaws
Lum contact move VS Air Balloon Flame Body -> burn -> user's Lum -> Air Balloon pops
Scald -> burn -> Synchronize -> target's Lum
Mold Breaker Nuzzle VS Limber Lum -> paralysis -> target's Lum
Mold Breaker Nuzzle VS Synchronize Lum -> paralysis -> Synchronize -> target's Lum
Sitrus contact move VS Rocky Helmet Rough Skin -> Rough Skin -> user's Sitrus -> Rocky Helmet
```

1. Most "attack is successful" effects
2. Additional effects
3. Berries can trigger here (target's status curing Berries)
4. Grudge/Rage
5. Most Ability effects
6. Berries can trigger here (user's Berries that check HP, user's status curing Berries)
7. Item effects that trigger when the target is hit by the attack (including Air Balloon and Sticky Barb)
8. Berries can trigger here (user's Berries that check HP)
9. The target faints if it has 0 HP
10. Destiny Bond if the target fainted
11. Enigma Berry
12. Recoil damage/Struggle
13. Berries can trigger here (user's Berries that check HP)
14. The target thaws out if attack is a Fire-type attack, Scald, Steam Eruption
15. Knock Off, Rapid Spin, Bug Bite/Pluck, Thief/Covet effects (only if the user has not fainted, and even if the target fainted), Dragon Tail/Circle Throw, SmellingSalt/Wake-Up Slap effects (only if the user has not fainted)
16. Berries can trigger here (target's Berries that check HP)


After all hits of a successful attack have resolved (Ability and item effects mentioned below don't occur if the attacker has Sheer Force and the attack has an additional effect):

    Color Change
    Kee and Maranga Berries
    Red Card (only if attacker and item's holder haven't fainted)
    Eject Button message (only if item's holder hasn't fainted; if more than one target's Eject Button could trigger, only the first such target in turn order triggers).
    Shell Bell/Life Orb (only if attacker hasn't fainted)
    Eject Button effect (only if item's holder hasn't fainted). If effect happens, the U-turn message/effect doesn't happen.
    U-turn message (only if attacker hasn't fainted)
    Pickpocket (only if the Ability's bearer hasn't fainted)
    U-turn effect (only if attacker hasn't fainted)

**Comment by Zarel** [comment]
Your previous code didn't use `.lastEffect`. Can you move back to that for now?

**Comment by QuiteQuiet** [comment]
I already did in https://github.com/Zarel/Pokemon-Showdown/pull/4045/commits/9701abef276cfcc85b1066042bee09a7d9bc2097

**Comment by Zarel** [comment]
I think the code as it currently stands is an improvement upon the previous situation. @Marty-D, LGTM?

**Comment by Marty-D** [comment]
Magic Room isn't the only thing that stops item usage, see also https://github.com/Zarel/Pokemon-Showdown/blob/master/sim/pokemon.js#L375-L377.

**Comment by QuiteQuiet** [comment]
Does a Pokemon under the effects of Embargo that switches out eat its berry? I think I've covered the other cases now, but I'm not sure about that one.

**Comment by Marty-D** [comment]
When it comes back in later, yeah. Not during the switch or anything.

**Comment by QuiteQuiet** [comment]
This should cover everything then.

**Comment by Marty-D** [comment]
I think Klutz needs an `onEnd` also?

**Comment by QuiteQuiet** [comment]
Oh, right. Did.

**Comment by Marty-D** [comment]
`onAfterEnd` -> `onEnd` for Klutz; `onAfterEnd` won't work for Abilities.

The last thing would be to include Berry Juice in all your checks since it's not actually a Berry.

**Comment by QuiteQuiet** [comment]
During the `onEnd` event Klutz is still active so it will always end with `pokemon.ignoreItem()` being true. The `onEnd` as a whole in that regard doesn't work for anything here, which is why the `onAfterEnd` exist at all.

**Comment by Marty-D** [comment]
I just tested it; using your code, Super Fang to Gastro Acid didn't activate the Berry, but when I change it to onEnd it works.

**Comment by QuiteQuiet** [comment]
I was testing with Super Fang to Simple Beam, which doesn't work with onEnd, but onAfterEnd does. I guess I could be doing both (or find out what's happening exactly), but just onEnd isn't enough.

**Comment by Marty-D** [comment]
Oh, weird. I'll have to look into that later. I assume ignoringAbility() is checked in hasAbility(), which lets Gastro Acid work.

**Comment by QuiteQuiet** [comment]
From what I can tell, the Gastro Acid volatile is applied and then the ability `onEnd` runs. onEnd checks `ignoringItem()`, which does `this.hasAbility('klutz')`, and hasAbility checks for Gastro Acid.

On the other hand, when an ability ends in `setAbility` the onEnd event runs before the ability is replaced, so in this case Klutz is still active during the onEnd event.

Moving https://github.com/Zarel/Pokemon-Showdown/blob/master/sim/pokemon.js#L1032 to the line above the `singleEvent('End')` call seems like a viable alternative alongside only the onEnd for Klutz, but I'm not sure if that would cause any unintended side-effects (the event is already called based on the old ability and ability data)

**Comment by Marty-D** [comment]
Hmm, I sprinkled exceptions to `AfterEnd` all over `singleEvent()` and nothing helped. I'm missing something here; any ideas, @Zarel?

**Comment by Zarel** [comment]
I'm confused. Can you describe the problem precisely?

**Comment by Marty-D** [comment]
Using `onEnd` for Klutz to make Berries activate immediately only works for the case where Gastro Acid stops Klutz from working, and `onAfterEnd` works for cases where Klutz gets replaced by some other Ability. How to make one of said events work for all cases?

**Comment by Zarel** [comment]
I would add a new item event, `StartUpdate` and trigger it before `Start` for items, as well as after Klutz. I would probably trigger it from `onAfterEnd` for Klutz, and figure out why it's not being triggered by Gastro Acid, and fix that (it's probably just missing a manual trigger).

I'd also consider this sort of change outside the scope of a single pull request, and split it into two.

**Comment by QuiteQuiet** [comment]
I believe I sorted out the Gastro Acid not properly triggering the `onAfterEnd` event for Klutz now, while still working for every other case too.

**Comment by Marty-D** [comment]
Oh, that's what I was missing, I never looked at Gastro Acid itself. Nice work, thanks! 👍

**Pull request merged**